contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,384 @@
+                if (curInfo.tsStart > tsCheckpointLast) {
+                    serviceCheckpoint = curInfo.tsStart;
+                }
+                // Calculate the liveness ratio in 1e18 value
+                uint256 ratio;
+                // If the checkpoint was called in the exactly same block, the ratio is zero
+                if (block.timestamp > serviceCheckpoint) {
+                    uint256 nonce = serviceNonces[i];
+                    ratio = ((nonce - curInfo.nonce) * 1e18) / (block.timestamp - serviceCheckpoint);
+                }
+
+                // Record the reward for the service if it has provided enough transactions
+                if (ratio >= livenessRatio) {
+                    // Calculate the reward up until now and record its value for the corresponding service
+                    uint256 reward = rewardsPerSecond * (block.timestamp - serviceCheckpoint);
+                    totalRewards += reward;
+                    eligibleServiceRewards[numServices] = reward;
+                    eligibleServiceIds[numServices] = curServiceId;
+                    ++numServices;
+                }
+            }
+        }
+    }
+
+    /// @dev Checkpoint to allocate rewards up until a current time.
+    /// @return All staking service Ids.
+    /// @return Number of staking eligible services.
+    /// @return Eligible service Ids.
+    /// @return Eligible service rewards.
+    function checkpoint() public returns (uint256[] memory, uint256, uint256[] memory, uint256[] memory) {
+        // Calculate staking rewards
+        (uint256 lastAvailableRewards, uint256 numServices, uint256 totalRewards,
+            uint256[] memory eligibleServiceIds, uint256[] memory eligibleServiceRewards,
+            uint256[] memory serviceIds, uint256[] memory serviceNonces) = _calculateStakingRewards();
+
+        // If available rewards are not zero, proceed with staking calculation
+        if (lastAvailableRewards > 0) {
+            // If total allocated rewards are not enough, adjust the reward value
+            if (totalRewards > lastAvailableRewards) {
+                // Traverse all the eligible services and adjust their rewards proportional to leftovers
+                totalRewards = 0;
+                for (uint256 i = 0; i < numServices; ++i) {
+                    // Calculate the updated reward
+                    uint256 updatedReward = (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
+                    // Add to the total updated reward
+                    totalRewards += updatedReward;
+                    // Add reward to the service overall reward
+                    uint256 curServiceId = eligibleServiceIds[i];
+                    mapServiceInfo[curServiceId].reward += updatedReward;
+                }
+
+                // If the reward adjustment happened to have small leftovers, add it to the last traversed service
```

Comment:
So last service in the list can receive a bit more?

Reply:
yes

Reply:
Is reverse loop more expensive? Makes sense to have the first staked service have this privilige.